gips/0065-subgraph-availability-manager.md

```diff
@@ -0,0 +1,123 @@
+
+The `SubgraphAvailabilityManager` contract will be governed by the council. While the contract itself will not be upgradeable, a new contract can be deployed, and the council can set the `RewardsManager` oracle address to the new deployment. The council will have the authority to add or remove oracles operators and update the `voteTimeLimit` timeframe.
+
+# Detailed Specification
+
+The number of oracle operators will be strictly enforced as five. Upon contract initialization, five valid accounts must be provided, and each oracle must maintain their index position, crucial for passing it as a parameter when casting votes.
+
+Vote aggregation will be separately tracked for positive or negative votes using the following data structures:
+
+```solidity
+mapping(uint256 currentNonce => mapping(bytes32 subgraphDeploymentID => uint256[NUM_ORACLES] timestamps)) public lastDenyVote;
+mapping(uint256 currentNonce => mapping(bytes32 subgraphDeploymentID => uint256[NUM_ORACLES] timestamps)) public lastAllowVote;
+```
+
+The initial mapping key corresponds to a public variable `currentNonce`. This nonce serves as a mechanism to invalidate all votes whenever a configuration is altered—be it updating the oracle operators list or modifying the `voteTimeLimit` parameter. The inner mapping stores the `subgraphDeploymentId` against the timestamp of each oracle's vote, using an array denoting the oracle index.
+
+The `executionThreshold` will be set to three during contract construction. Whenever an oracle casts a vote, the `SubgraphAvailabilityManager` contract will query the corresponding vote aggregation storage to verify if enough votes have been registered. Upon the third vote a transaction to `RewardsManager` is triggered to make the subgraph rewards state change.
+
+A `voteTimeLimit` parameter will be used to invalidate old votes, initially set to ten minutes. This duration is slightly longer than the current subgraph availability oracle configuration, which runs every five minutes, allowing time for oracles to cast their votes and accounting for potential delays in execution. This minimizes the need for frequent transactions. When comparing oracle votes, the `SubgraphAvailabilityManager` will use the `voteTimeLimit` to ensure the last recorded vote occurred within this timeframe. Votes older than the limit will not be considered valid, necessitating a new vote from the oracle if required. It's crucial to note that any adjustments to `voteTimeLimit` must always exceed the oracle's execution period to prevent the premature invalidation of votes before all oracles have participated. However, this timeframe should be kept as brief as possible, since an excessively long `voteTimeLimit` undermines its intended purpose. The council can modify this timeframe if desired but it must be set with consideration to the oracle’s iteration period to avoid repeated voting or expired votes.
```

Comment:
If `voteTimeLimit` expires without a consensus vote, are oracles are expected to start a new voting round? 

There is a scenario where a large spike in onchain gas price could deplete some of the oracle account balances, which will put them offline, and in turn force the other oracles to spend their remaining funds in doing repeated voting rounds, depleting their own funds in the process.

A backoff mechanism could help, but maybe the detailed software mechanism are outside the scope of this GIP?

Reply:
Indeed, if `voteTimeLimit` expires oracles will start a new voting round. This should be covered in `Responsibility and Integrity` section of the Charter where operators are required to have oracles running at all times. Part of having the oracle running is having the oracle wallet funded. Changes to the oracle's code are out of scope for this GIP but I'm not sure if we want to backoff since fix should come from offline oracles. However, I agree that depleting funds with no purpose isn't ideal. This situation should be minimized by having 5 oracles and requiring 3 votes, so even with 2 oracles offline, the system should remain operational.

Reply:
This could be avoided by having continuous voting. Instead of 10 minutes rounds, oracles can vote at any time. As soon as 3 vote "Deny", the subgraph is denied, and renabled as soon as 1 of the 3 vote "Allow". Non-voting oracles are considered a "Allow".  That would remove the need for repeated identical votes.

What's the reasoning behind voting in rounds?

Reply:
Sorry, not sure where I explained incorrectly but this is how it currently works, with the exception that it also requires 3 votes for an "Allow" transaction.  Oracles run every 5 minutes, each time they run they'll vote if they see that a subgraph state needs to change. The 10 minutes `voteTimeLimit` is a safety mechanism so as to not leave the system where one malicious vote could trigger a change in `RewardsManager`. For example if two oracles incorrectly vote to deny a subgraph and after that they correct themselves if we don't expire those old votes a new oracle with just one vote could change the state for that subgraph.

Reply:
I'm not sure I understand the reason behind the voting rounds. In this example, if two oracles incorrectly vote to deny, and they correct themselves minutes or hours later, couldn't they simply change their vote? I don't understand why we'd need the concept of a time limit to invalidate these initial votes.

Reply:
Oracles only cast votes when a change in rewards is needed. If two oracles wrongly vote to deny a subgraph and after that they correct themselves they wouldn't be changing their vote in the smart contract, they would just stop sending transactions.